### PR TITLE
Introduce ActorPreviewWidget (and other related changes).

### PR DESCRIPTION
**Pull request by pchote**
There are three parts to this pr:
1. Make the IRenderable structs properly immutable, and introduce a new type for prepared renderables.
2. Introduce the `ActorPreviewWidget` itself.
3. Replace the hacky `SpriteSequenceWidget` used by the color pickers with `ActorPreviewWidget`.

https://github.com/pchote/OpenRA/tree/editor provides a heavier testcase of `ActorPreviewWidget`.

### CODE REVIEW

**Comment by Phrohdoh** [comment]
Need to update the header years.

**Comment by pchote** [comment]
Good spotting. Fixed.

File: OpenRA.Game/Graphics/Renderable.cs [hunk 1]
```diff
@@ -17,23 +17,26 @@ namespace OpenRA.Graphics
...
 		IRenderable WithPalette(PaletteReference newPalette);
 		IRenderable WithZOffset(int newOffset);
 		IRenderable OffsetBy(WVec offset);
 		IRenderable AsDecoration();
 
-		void BeforeRender(WorldRenderer wr);
+		IFinalizedRenderable PrepareRender(WorldRenderer wr);
+	}
+
+	public interface IFinalizedRenderable
```
Comment (penev92) [inline OpenRA.Game/Graphics/Renderable.cs]: This and `IRenderable` would be better off in a separate interface file.

Comment (pchote) [inline OpenRA.Game/Graphics/Renderable.cs]: This *is* the separate interface file, but I agree. I'll split `SpriteRenderable` out into its own file instead.

**Comment by Mailaender** [comment]
> https://github.com/pchote/OpenRA/tree/editor provides a heavier testcase of `ActorPreviewWidget`.

Can you rebase onto the latest revision of #4496 to get tooltip support, because this change breaks things and I don't know for which actors:

image

It crashes when you try to place those.

**Comment by pchote** [comment]
I'll leave that to you, as it is not really part of this pull request.

**Comment by penev92** [comment]
Rebasing onto that PR would make this depend on that, while it's supposed to be the other way around.

**Comment by Mailaender** [comment]
Sorry, :-1: then. This does not work with any infantry actor and will therefor break #4496.

**Comment by pchote** [comment]
You're complaining about changes that aren't in this pr?

**Comment by Mailaender** [comment]
I fear that I will have to rebase on this, things will break and I won't be able to fix it.

**Comment by pchote** [comment]
Could you please elaborate on what specific changes you are requesting for this PR?

**Comment by Mailaender** [comment]
Fix the following crashes that do not appear with the ActorPreviewWidget in #4496:

```
Exception of type `System.ArgumentNullException`: Argument cannot be null.
Parameter name: source
  at System.Linq.Check.SourceAndSelector (System.Object source, System.Object selector) [0x00017] in /home/abuild/rpmbuild/BUILD/mono-3.10.0/mcs/class/System.Core/System.Linq/Check.cs:73 
  at System.Linq.Enumerable.Select[KeyValuePair`2,IRenderable] (IEnumerable`1 source, System.Func`2 selector) [0x00000] in /home/abuild/rpmbuild/BUILD/mono-3.10.0/mcs/class/System.Core/System.Linq/Enumerable.cs:2262 
  at OpenRA.Mods.Common.Traits.Bridge.TemplateRenderables (OpenRA.Graphics.WorldRenderer wr, OpenRA.Graphics.PaletteReference palette, UInt16 template) [0x0004b] in OpenRA.Mods.Common/Traits/Buildings/Bridge.cs:178 
  at OpenRA.Mods.Common.Traits.Bridge.Render (OpenRA.Actor self, OpenRA.Graphics.WorldRenderer wr) [0x00059] in OpenRA.Mods.Common/Traits/Buildings/Bridge.cs:192 
  at OpenRA.Actor+<Renderables>c__Iterator0.MoveNext () [0x00074] in OpenRA.Game/Actor.cs:136 
  at System.Linq.Enumerable+<CreateSelectManyIterator>c__Iterator12`2[OpenRA.Actor,OpenRA.Graphics.IRenderable].MoveNext () [0x0009a] in /home/abuild/rpmbuild/BUILD/mono-3.10.0/mcs/class/System.Core/System.Linq/Enumerable.cs:2304 
  at System.Linq.Enumerable+<CreateConcatIterator>c__Iterator1`1[OpenRA.Graphics.IRenderable].MoveNext () [0x0005d] in /home/abuild/rpmbuild/BUILD/mono-3.10.0/mcs/class/System.Core/System.Linq/Enumerable.cs:596 
  at System.Linq.Enumerable.ToArray[IRenderable] (IEnumerable`1 source) [0x0007c] in /home/abuild/rpmbuild/BUILD/mono-3.10.0/mcs/class/System.Core/System.Linq/Enumerable.cs:2900 
  at System.Linq.QuickSort`1[OpenRA.Graphics.IRenderable]..ctor (IEnumerable`1 source, System.Linq.SortContext`1 context) [0x00006] in /home/abuild/rpmbuild/BUILD/mono-3.10.0/mcs/class/System.Core/System.Linq/QuickSort.cs:44 
  at System.Linq.QuickSort`1+<Sort>c__Iterator0[OpenRA.Graphics.IRenderable].MoveNext () [0x00021] in /home/abuild/rpmbuild/BUILD/mono-3.10.0/mcs/class/System.Core/System.Linq/QuickSort.cs:73 
  at System.Linq.Enumerable+<CreateConcatIterator>c__Iterator1`1[OpenRA.Graphics.IRenderable].MoveNext () [0x0005d] in /home/abuild/rpmbuild/BUILD/mono-3.10.0/mcs/class/System.Core/System.Linq/Enumerable.cs:596 
  at System.Linq.Enumerable+<CreateSelectIterator>c__Iterator10`2[OpenRA.Graphics.IRenderable,OpenRA.Graphics.IFinalizedRenderable].MoveNext () [0x00059] in /home/abuild/rpmbuild/BUILD/mono-3.10.0/mcs/class/System.Core/System.Linq/Enumerable.cs:2270 
  at System.Collections.Generic.List`1[OpenRA.Graphics.IFinalizedRenderable].AddEnumerable (IEnumerable`1 enumerable) [0x00013] in /home/abuild/rpmbuild/BUILD/mono-3.10.0/mcs/class/corlib/System.Collections.Generic/List.cs:146 
  at System.Collections.Generic.List`1[OpenRA.Graphics.IFinalizedRenderable]..ctor (IEnumerable`1 collection) [0x0002f] in /home/abuild/rpmbuild/BUILD/mono-3.10.0/mcs/class/corlib/System.Collections.Generic/List.cs:69 
  at System.Linq.Enumerable.ToList[IFinalizedRenderable] (IEnumerable`1 source) [0x00006] in /home/abuild/rpmbuild/BUILD/mono-3.10.0/mcs/class/System.Core/System.Linq/Enumerable.cs:2952 
  at OpenRA.Graphics.WorldRenderer.GenerateRenderables () [0x0011a] in OpenRA.Game/Graphics/WorldRenderer.cs:105 
  at OpenRA.Graphics.WorldRenderer.Draw () [0x00032] in OpenRA.Game/Graphics/WorldRenderer.cs:119 
  at OpenRA.Game.RenderTick () [0x00070] in OpenRA.Game/Game.cs:503 
  at OpenRA.Game.Loop () [0x0013e] in OpenRA.Game/Game.cs:625 
  at OpenRA.Game.Run () [0x00042] in OpenRA.Game/Game.cs:645 
  at OpenRA.Program.Run (System.String[] args) [0x00011] in OpenRA.Game/Support/Program.cs:111 
  at OpenRA.Program.Main (System.String[] args) [0x00050] in OpenRA.Game/Support/Program.cs:39
```

It seems I was wrong. Infantry works. It is the bridge actors that don't work properly.

**Comment by pchote** [comment]
I'm unable to reproduce that on my editor branch, because aren't visible in the actors tab.  How can I trigger this?

In any case, the bridge actors are generated at runtime by the engine, so shouldn't be placed manually (but that doesn't mean that they should be allowed to crash).

**Comment by Mailaender** [comment]
Look for the invisible tiny squares in the actor selector at the position I indicated in my screenshot.

**Comment by pchote** [comment]
That crash occurs when a bridge is placed because `footprint` is null (which is because it was placed by hand instead of auto-generated by the engine).  It's not related to this PR.

You don't see this issue in the current version of #4496 because it throws the exception inside a try/catch loop that discards the actor from the selector list.

File: OpenRA.Game/Graphics/WorldRenderer.cs [hunk 2]
```diff
@@ -146,21 +143,19 @@ public void Draw()
...
 			var overlayRenderables = World.Selection.Actors.Where(a => !a.Destroyed)
 				.SelectMany(a => a.TraitsImplementing<IPostRenderSelection>())
-				.SelectMany(t => t.RenderAfterWorld(this))
-				.ToList();
+				.SelectMany(t => t.RenderAfterWorld(this));
 
 			Game.Renderer.WorldVoxelRenderer.BeginFrame();
-			for (var i = 0; i < overlayRenderables.Count; i++)
-				overlayRenderables[i].BeforeRender(this);
+			var finalOverlayRenderables = overlayRenderables.Select(r => r.PrepareRender(this));
```
Comment (Mailaender) [inline OpenRA.Game/Graphics/WorldRenderer.cs]: You are shifting `BeforeRender` and `PrepareRender` here. Is that intentional?

**Comment by Mailaender** [comment]
Color pickers still work. :white_check_mark:

File: OpenRA.Game/Graphics/WorldRenderer.cs (hunk 2 above)
Comment (pchote) [inline OpenRA.Game/Graphics/WorldRenderer.cs]: Yes.

**Comment by penev92** [comment]
Well I did quite a bit of ingame testing... Found nothing wrong.
Also nothing in the code struck me as ugly or dangerous. :+1: 
The titan was a nice touch ;)

File: OpenRA.Mods.Common/Widgets/ActorPreviewWidget.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+			IdealPreviewSize = int2.Zero;
+
+			var r = preview
+				.SelectMany(p => p.Render(worldRenderer, WPos.Zero))
+				.OrderBy(WorldRenderer.RenderableScreenZPositionComparisonKey)
+				.Select(rr => rr.PrepareRender(worldRenderer));
+
+			if (r.Any())
+			{
+				var b = r.First().ScreenBounds(worldRenderer);
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Widgets/ActorPreviewWidget.cs]: `var b = r` ... I should have complained about this cryptic code earlier.

**Pull request merged**